src/gui/properties/speedplotview.cpp

```diff
@@ -66,67 +67,138 @@ SpeedPlotView::SpeedPlotView(QWidget *parent)
     greenPen.setStyle(Qt::DotLine);
     m_properties[TRACKER_UP] = GraphProperties(tr("Tracker Upload"), bluePen);
     m_properties[TRACKER_DOWN] = GraphProperties(tr("Tracker Download"), greenPen);
+
+    initBuffers(MIN5_BUF_SIZE, m_xData5Min, m_yData5Min);
+    initBuffers(MIN30_BUF_SIZE, m_xData30Min, m_yData30Min);
+    initBuffers(HOUR6_BUF_SIZE, m_xData6Hour, m_yData6Hour);
+}
+
+void SpeedPlotView::initBuffers(int capacity, boost::circular_buffer<uint> &xBuffer,
+                                QMap<SpeedPlotView::GraphID, boost::circular_buffer<int> > &yBuffers)
+{
+    xBuffer = boost::circular_buffer<uint>(capacity);
+    yBuffers[UP] = boost::circular_buffer<int>(capacity);
+    yBuffers[DOWN] = boost::circular_buffer<int>(capacity);
+    yBuffers[PAYLOAD_UP] = boost::circular_buffer<int>(capacity);
+    yBuffers[PAYLOAD_DOWN] = boost::circular_buffer<int>(capacity);
+    yBuffers[OVERHEAD_UP] = boost::circular_buffer<int>(capacity);
+    yBuffers[OVERHEAD_DOWN] = boost::circular_buffer<int>(capacity);
+    yBuffers[DHT_UP] = boost::circular_buffer<int>(capacity);
+    yBuffers[DHT_DOWN] = boost::circular_buffer<int>(capacity);
+    yBuffers[TRACKER_UP] = boost::circular_buffer<int>(capacity);
+    yBuffers[TRACKER_DOWN] = boost::circular_buffer<int>(capacity);
 }
 
 void SpeedPlotView::setGraphEnable(GraphID id, bool enable)
 {
     m_properties[id].m_enable = enable;
+    this->viewport()->update();
 }
 
-void SpeedPlotView::pushXPoint(double x)
+void SpeedPlotView::pushXPoint(uint x)
 {
-    while (m_xData.size() >= m_maxCapacity)
-        m_xData.pop_front();
+    ++m_counter30Min;
+    m_counter30Min %= 3;
+    ++m_counter6Hour;
+    m_counter6Hour %= 6;
 
-    m_xData.append(x);
+    m_xData5Min.push_back(x);
+
+    if (m_counter30Min == 0)
+        m_xData30Min.push_back(x);
+
+    if (m_counter6Hour == 0)
+        m_xData6Hour.push_back(x);
```

Comment:
You are saving 1 point of each 6 instead of doing the average. It's ok but you can improve accuracy in this way:
Create one var to store the sum, m_sum6Hour
```
if (m_counter6Hour == 0)
   m_xData6Hour.push_back(m_sum6Hour / 6);
   m_sum6Hour = 0;
else
   m_sum6Hour += x;
```